protocol-select/README.md

```diff
@@ -0,0 +1,541 @@
+            - [0-RTT](#0-rtt)
+        - [Stream Protocol Negotiation](#stream-protocol-negotiation)
+            - [Initiator](#initiator)
+            - [Listener](#listener)
+    - [Transitioning from [Multistream Select]](#transitioning-from-multistream-select)
+        - [Multiphase Rollout](#multiphase-rollout)
+            - [Phase 1](#phase-1)
+            - [Phase 2](#phase-2)
+            - [Phase 3](#phase-3)
+        - [Additional Rollout Mechanisms](#additional-rollout-mechanisms)
+        - [Heuristics](#heuristics)
+            - [TCP](#tcp)
+            - [QUIC](#quic)
+            - [Protocol Differentiation](#protocol-differentiation)
+    - [Protocol Specification](#protocol-specification)
+        - [Protocol Evolution](#protocol-evolution)
+            - [Non-Breaking Changes](#non-breaking-changes)
+            - [Breaking Changes](#breaking-changes)
+    - [Extensions](#extensions)
+        - [Protocol IDs](#protocol-ids)
+    - [FAQ](#faq)
+
+## Introduction
+
+_Protocol Select_ is a protocol negotiation protocol. It is aimed at negotiating
+libp2p protocols on connections and streams. It replaces the _[Multistream
+Select]_ protocol.
+
+### Improvements over _[Multistream Select]_
+
+- **Downgrade attacks** and **censorship resistance**
+
+  Given that **[Multistream Select]** negotiates a connection's security
+  protocol unencrypted and unauthenticated it is prone to [downgrade attack]s.
+  In addition, a man-in-the-middle can detect that a given connection is used
+  to carry libp2p traffic, allowing attackers to censor such connections.
+
+  **Protocol Select** is combined with a change to the Multiaddr format,
+  advertising the secure channel protocol through the latter instead of
+  negotiating them in-band. Thus [Downgrade attack]s are no longer possible at
+  the protocol negotiation level and a man-in-the-middle can no longer detect
+  a connection being used for libp2p traffic through the negotation process.
+
+- **Connection establishment**
+
+  In addition to making us vulnerable to downgrade attacks, negotiating the
+  security protocol takes one round-trip in the common case with **[Multistream
+  Select]**. On top of that negotiating a stream multiplexer (on TCP) takes
+  another round-trip.
+
+  **Protocol Select** on the other hand depends on security protocols being
+  advertised, thereby eliminating the need for negotiating them. For optimized
+  implementations, stream muxer negotiation will take zero round-trips for the
+  client (depending on the details of the cryptographic handshake protocol). In
+  that case, the client will be able to immediately open a stream after
+  completing the cryptographic handshake. In addition the protocol supports
+  zero-round-trip optimistic stream protocol negotiation when proposing a single
+  protocol.
+
+- **Data schema**
+
+  The **[Multistream Select]** protocol is defined as a plaintext protocol
```

Comment:
This isn't really correct. I'd instead say that multistream used a bespoke format that didn't clearly delineate between different atomic "messages".

That was the issue here, as far as I can tell.

Reply:
I applied your suggestion in 4ccea37.